### PR TITLE
feat: make statement result export async

### PR DESCRIPTION
**Pull request by kgopal492**
- Make export async

- Allow entire statement results to be exported to exporters (up to exporter limits)
  - Add `get_csv_iter` as an abstract method in `BaseReader` to read query results in batches
  - `read_csv` method in `BaseReader` now unpacks results from `get_csv_iter`
  - `gspread_exporter` is updated to suppport reading entire results past `STORE_MAX_READ_SIZE` (up to gspread limit)

### CODE REVIEW

File: querybook/server/logic/result_store.py [hunk 1]
```diff
@@ -58,13 +58,13 @@ def delete_key_value_store(key, commit=True, session=None):
             session.commit()
 
 
-def string_to_csv(raw_csv_str: str) -> List[List[str]]:
+def get_csv_reader(raw_csv_str: str) -> Generator[List[List[str]], None, None]:
```
Comment (czgu) [inline querybook/server/logic/result_store.py:61]: str_to_csv_iter

File: querybook/server/tasks/export_query_execution.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    socketio.emit(
+        "export_status_info",
+        {
+            "task_id": self.request.id,
+            "statement_id": statement_execution_id,
+            "status": QueryExecutionExportStatus.RUNNING.value,
+        },
+        namespace=DATA_DOC_NAMESPACE,
+        room=socket_id,
+        broadcast=True,
```
Comment (czgu) [inline querybook/server/tasks/export_query_execution.py:29]: no broadcast needed?

File: querybook/webapp/components/DataDocStatementExecutionBar/ResultExportSuccessToast.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        <div className="flex-center mv24">
+            <Title size={5} title={title} />
+            <Link to={url} newTab>
+                View Export
+            </Link>
+        </div>
+    </Modal>
+);
+
+export const ResultExportSuccessToast = (
```
Comment (czgu) [inline querybook/webapp/components/DataDocStatementExecutionBar/ResultExportSuccessToast.tsx:25]: SuccessModal?

File: querybook/webapp/lib/data-doc/datadoc-socketio.ts [hunk 4]
```diff
@@ -70,6 +73,12 @@ export interface IDataDocSocketEvent {
         ) => any
     >;
 
+    receiveExportStatus?: IDataDocSocketEventPromise<
```
Comment (czgu) [inline querybook/webapp/lib/data-doc/datadoc-socketio.ts:76]: don't bundle export query results with datadoc socket since export can happen without a datadoc being opened

File: querybook/server/lib/export/base_exporter.py [hunk 5]
```diff
@@ -64,24 +64,41 @@ def export(self, statement_execution_id: int, uid: int, **options) -> str:
...
-                if raw:
-                    result = "\n".join(
-                        reader.read_lines(number_of_lines=number_of_lines)
-                    )
-                else:
-                    result = reader.read_csv(number_of_lines=number_of_lines)
-                return result
+            with GenericReader(
+                statement_execution.result_path,
+                max_read_size=number_of_lines,  # override max_read_size in some readers
```
[the remark this answers is not in the document]
Comment (kgopal492) [reply querybook/server/lib/export/base_exporter.py:99]: changed to `max_read_size=None` => this will allow reader to only be limited by `number_of_lines` passed

File: querybook/server/lib/export/base_exporter.py [hunk 6]
```diff
@@ -86,6 +89,29 @@ def _get_statement_execution_columns_len(
         return 0
 
     def _get_statement_execution_result(
+        self,
+        statement_execution_id: int,
+        raw: bool = False,  # If raw, return unparsed csv text
+        number_of_lines: int = 2001,
+    ):
+        """Warning: This function will be deprecated shortly as we move towards exporting full
```
Comment (czgu) [inline querybook/server/lib/export/base_exporter.py:97]: do something like `LOG.error('_get_statement_execution_result will be deprecated soon due to ...')`

File: querybook/server/lib/export/exporters/gspread_exporter.py [hunk 7]
```diff
@@ -105,17 +106,20 @@ def export_form(self):
...
+            statement_execution_id, session=session,
         )
         start_cell_coord = worksheet_coord_to_coord(start_cell)
         sheet_start_coord = worksheet_coord_to_coord("A1")
 
         column_offset = start_cell_coord[0] - sheet_start_coord[0]
         row_offset = start_cell_coord[1] - sheet_start_coord[1]
 
-        total_column_cells = result_columns_len + column_offset
+        total_column_cells = max(result_columns_len + column_offset, 26)
```
Comment (czgu) [inline querybook/server/lib/export/exporters/gspread_exporter.py:122]: add a comment about 26, or move 26 to some constant like DEFAULT_GSPREAD_NUM_COLS

File: querybook/server/datasources/query_execution.py [hunk 8]
```diff
@@ -370,18 +370,44 @@ def export_statement_execution_result(
         api_assert(valid, "Invalid exporter params, reason: " + reason)
 
     task = export_query_execution_task.apply_async(
-        args=[
-            exporter.exporter_name,
-            statement_execution_id,
-            current_user.id,
-            originator,
-        ],
+        args=[exporter.exporter_name, statement_execution_id, current_user.id,],
```
Comment (czgu) [inline querybook/server/datasources/query_execution.py:373]: instead of having exporter_params passed as kwargs, can we have a parameter called exporter_params instead?

File: querybook/tests/test_lib/test_export/test_exporters/test_gspread_exporter.py [hunk 9]
```diff
@@ -57,7 +55,7 @@ def test_get_no_columns(self):
         # edge case - 0 columns (ie no statement results) shouldn't break this function
         self.statement_columns_len_mock.return_value = 0
         exporter = GoogleSheetsExporter({})
-        self.assertEqual(exporter._get_max_rows(1), MAX_SHEETS_CELLS)
+        self.assertEqual(exporter._get_max_rows(1), MAX_SHEETS_CELLS // 26)
```
Comment (czgu) [inline querybook/tests/test_lib/test_export/test_exporters/test_gspread_exporter.py:58]: i think in this case we can just assert its not zero and there is no exception, there is no need to check the divide by 26 logic (unless you want to have a test case specifically for that)

File: querybook/webapp/components/DataDocStatementExecutionBar/ResultExportDropdown.tsx [hunk 10]
```diff
@@ -113,16 +118,49 @@ export const ResultExportDropdown: React.FunctionComponent<IProps> = ({
...
-                    formData
-                )
+            await getExporterAuthentication(exporter);
+            const { data: taskId } = await StatementResource.export(
+                statementId,
+                exporter.name,
+                formData
+            );
+
+            const pollExportPromise = new Promise((res, rej) => {
```
Comment (czgu) [inline querybook/webapp/components/DataDocStatementExecutionBar/ResultExportDropdown.tsx:132]: can you extract the promise logic here:

```
new Promise((res, rej) => {
                const poll = setInterval(async () => {
                    const { data } = await StatementResource.pollExportTask(
                        taskId
                    );
                    const { status } = data;
                    if (status === QueryExecutionExportStatus.ERROR) {
                        clearInterval(poll);
                        rej(data.message ?? 'unknown error');
                    } else if (status === QueryExecutionExportStatus.DONE) {
                        clearInterval(poll);
                        res(data);
                    }
                }, 5000);
            })
```
to lib/result-export.ts

File: querybook/webapp/components/DataDocStatementExecutionBar/ResultExportDropdown.tsx [hunk 11]
```diff
@@ -113,16 +118,49 @@ export const ResultExportDropdown: React.FunctionComponent<IProps> = ({
...
+                        clearInterval(poll);
+                        res(data);
+                    }
+                }, 5000);
+            });
+            return toast.promise(
+                pollExportPromise,
+                {
+                    loading: 'Exporting, please wait',
+                    success: (data: IQueryExecutionExportStatusInfo) =>
```
Comment (czgu) [inline querybook/webapp/components/DataDocStatementExecutionBar/ResultExportDropdown.tsx:151]: success: ResultExportSuccessToast,

File: querybook/webapp/lib/result-export.ts [hunk 12]
```diff
@@ -36,3 +39,18 @@ export function getExporterAuthentication(
...
 }
+
+export const pollExporterTaskPromise = (taskId: string) =>
+    new Promise((res, rej) => {
+        const poll = setInterval(async () => {
+            const { data } = await StatementResource.pollExportTask(taskId);
+            const { status } = data;
+            if (status === QueryExecutionExportStatus.ERROR) {
+                clearInterval(poll);
+                rej(data.message ?? 'unknown error');
```
Comment (czgu) [inline querybook/webapp/lib/result-export.ts:50]: rej(new Error(data.message ?? 'unknown error'));

**Review by czgu — APPROVED** [review]

**Review by czgu — APPROVED** [review]
